File: src/main/java/sk/stuba/pks/old/service/PacketSender.java
package sk.stuba.pks.old.service;

import lombok.Getter;
import lombok.Setter;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Service;
import sk.stuba.pks.old.dto.Packet;
import sk.stuba.pks.old.service.states.session.Context;

import java.net.DatagramPacket;
import java.net.DatagramSocket;
import java.net.InetAddress;
import java.util.Deque;
import java.util.concurrent.ConcurrentLinkedDeque;

@Service
public class PacketSender implements Sender {

    private final DatagramSocket socket;

    @Getter
    @Setter
    private Deque<Packet> packetQueue = new ConcurrentLinkedDeque<>();
    public PacketSender(DatagramSocket socket) {
        this.socket = socket;
    }

    @Autowired
    private Context context;

    @Override
    public void sendPacket(Packet packet, String serverAddress, int serverPort) {
        try {
            byte[] data = packet.getBytes();
            InetAddress address = InetAddress.getByName(serverAddress);
            DatagramPacket datagramPacket = new DatagramPacket(data, data.length, address, serverPort);
            socket.send(datagramPacket);
        } catch (Exception e) {
            e.printStackTrace();
        }
    }

    public void startSendingPackets() {
        Runnable task = () -> {
            while (true) {
                // ensure thread is not interrupted
                if (Thread.currentThread().isInterrupted()) {
                    System.out.println("Thread interrupted");
                    return;
                }
                Packet packet = packetQueue.poll();
                if (packet == null) {
                    continue;
                }
                sendPacket(packet, context.getRemoteIp(), context.getRemotePort());
            }
        };
        Thread thread = new Thread(task);
        thread.start();
    }
}

File: src/main/java/sk/stuba/pks/library/PacketSender.kt
package sk.stuba.pks.library

import io.ktor.network.sockets.*
import io.ktor.utils.io.core.*
import kotlinx.coroutines.delay
import kotlinx.coroutines.withTimeout
import sk.stuba.pks.old.dto.Packet
import java.net.BindException
import java.net.SocketException
import java.util.*
import java.util.concurrent.ConcurrentLinkedDeque

class PacketSender(
    private val socket: BoundDatagramSocket,
    private val serverAddress: String,
    private val serverPort: Int,
    private val connectionTimeoutMs: Long,
    private val reconnectEveryMs: Long,
) {
    private val packetQueue: Deque<Packet> = ConcurrentLinkedDeque()

    suspend fun startSendingPackets() {
        while (true) {
            val packet: Packet? = packetQueue.poll()
            packet?.let {
                sendPacket(it)
            }
        }
    }

    private var sent = 0

    fun addPacket(packet: Packet) {
        packetQueue.add(packet)
    }

    fun addPacketToBeginning(packet: Packet) {
        packetQueue.addFirst(packet)
    }

    suspend fun sendPacket(packet: Packet) {
        sent++
        var isSent = false
        val data = packet.bytes
        withTimeout(connectionTimeoutMs) {
            while (!isSent) {
                try {
                    val addrs = InetSocketAddress(serverAddress, serverPort)
                    val byteReadPacket = ByteReadPacket(data)
                    val datagramPacket = Datagram(byteReadPacket, addrs)
                    socket.send(datagramPacket)
                    isSent = true
                } catch (e: BindException) {
                    println("Cant bind, retrying")
                    delay(reconnectEveryMs)
                } catch (e: SocketException) {
                    println("Cant bind, retrying")
                    delay(reconnectEveryMs)
                }
            }
            println("Sent packet $sent")
        }
    }

    fun clearQueue() {
        packetQueue.clear()
    }
}
